### ISSUE TITLE
Cloud code causes high CPU usage when not authenitcated

### ISSUE THREAD

**Issue by bikerp** [GoogleCloudPlatform/cloud-code-intellij#2875]
### Actual Behavior
Cloud code causes high CPU usage when not authenitcated. Once I authenticate using `gcloud auth login` then the CPU usage drops to normal.

### Additional Information

- IDE type IntelliJ
- IDE version: IntelliJ IDEA 2020.3.1 (Ultimate Edition)
- Cloud Code version 21.1.1-202
- Cloud SDK 
  - Are you allowing the plugin to manage the Cloud SDK:  No
  - Version of the Cloud SDK: 324.0.0
- Skaffold version (If you are using Kubernetes features; Settings > Cloud Code > Kubernetes): 1.17.2
- Operating System: Ubuntu 20.04.1

**Comment by etanshaul** [comment]
@bikerp thanks for reporting. We will investigate this.

> Are you allowing the plugin to manage the Cloud SDK: No

To confirm, you have your own installation of the Cloud SDK on your system that you are pointing Cloud Code to (Under Settings > Cloud Code > Cloud SDK > "Use Custom Local Installation") ?

**Comment by etanshaul** [comment]
Kamino cloned this issue to [GoogleCloudPlatform/cloud-code-intellij-internal](https://github.com/GoogleCloudPlatform/cloud-code-intellij-internal/issues/3388)

**Comment by bikerp** [comment]
> @bikerp thanks for reporting. We will investigate this.
> 
> > Are you allowing the plugin to manage the Cloud SDK: No
> 
> To confirm, you have your own installation of the Cloud SDK on your system that you are pointing Cloud Code to (Under Settings > Cloud Code > Cloud SDK > "Use Custom Local Installation") ?

Yes, I have my own instalation of Google Cloud SDK in /usr/lib/google-cloud-sdk folder

**Comment by ivanporty** [comment]
@bikerp do you use any of the Cloud Code features that require login, i.e. Cloud Run, Secrets, APIs, etc.? What would be the reason you only login with `gcloud auth login` on the command line and do you ever need to be logged in to GCP?

**Comment by bikerp** [comment]
> @bikerp do you use any of the Cloud Code features that require login, i.e. Cloud Run, Secrets, APIs, etc.? What would be the reason you only login with `gcloud auth login` on the command line and do you ever need to be logged in to GCP?

I mostly utilize code completion for Kubernetes manifests. I used Skaffold for deployment into local minikube in another project. I login with `gcloud auth login` because I need to interact with GCP GKE using kubectl CLI and I have to be logged in using corporate login (OIDC). I observed that the described behavior occurs when the IntelliJ project has Dockerfile and docker-compose file.

**Comment by etanshaul** [comment]
>  I observed that the described behavior occurs when the IntelliJ project has Dockerfile and docker-compose file.

That is an interesting factor. I wouldn't have expected that to correlate to cpu spikes related to login. We'll look into it.

**Comment by ivanporty** [comment]
> Cloud code causes high CPU usage when not authenitcated.

Hi @bikerp we are evaluating the issue. Could you please add more details on what `high CPU usage` means, how high, and how did you measure before and after GCP login?

**Comment by ivanporty** [comment]
> I observed that the described behavior occurs when the IntelliJ project has Dockerfile and docker-compose file.

And if you don't mind giving us a bit more detail on your setup, did you observe high CPU only if a project has Dockerfile and docker-compose file? How Dockerfile, docker-compose and GCP login state depend on each other, i.e. removing any of them from the equation removes high CPU issue?

**Comment by bikerp** [comment]
I checked the processes and this is what causes high CPU usage ` /usr/bin/python -S /usr/lib/google-cloud-sdk/lib/gcloud.py auth describe <my login> --format json --quiet` It is a subprocess of IntelliJ IDEA.
If I run this command manually I get this response:
```Reauthentication required.
ERROR: (gcloud.auth.describe) Please run:

  $ gcloud auth login

to complete reauthentication.```

**Comment by ivanporty** [comment]
Thanks @bikerp that is an excellent diagnostics, very helpful for us! We will work on this to reduce the unnecessary operations.

**Comment by ivanporty** [comment]
@bikerp We are not seeing this process when we are not logged in with Cloud SDK. Do you recall any other details on seeing this - did you sign out just before seeing this, used something with Cloud SDK, and what is the version of your Cloud SDK - it seems you're using your own Cloud SDK rather than managed version, correct?

**Comment by bikerp** [comment]
@ivanporty Yes I custom Cloud SDK 326.0.0. The described behavior occurs even when I turn on my computer and run IntelliJ. No prior interaction with Cloud SDK whatsoever. It occurs even on Angular projects with no Dockerfile or any Docker/Kubernetes manifests.

**Comment by etanshaul** [comment]
@ivanporty did you test on linux by any chance? I wonder if something `gcloud auth describe` command isn't gracefully exiting. @bikerp 's investigation _seems_ to suggest that the process sticks around

**Comment by bikerp** [comment]
I have noticed another process with high CPU usage `/usr/bin/python -S /usr/lib/google-cloud-sdk/lib/gcloud.py config config-helper --format json --quiet`. When trying to run manually then it says:
```
Reauthentication required.
ERROR: (gcloud.config.config-helper) Please run:

  $ gcloud auth login

to complete reauthentication.
```

**Comment by ivanporty** [comment]
Thank you for more details @bikerp! What happens if you run the same commands yourself from a terminal (`gcloud config config-helper --format json --quiet` or `gcloud auth describe <my login> --format json --quiet` - do they result in the high CPU usage, do they terminate right away?

**Comment by bikerp** [comment]
@ivanporty None of these commands cause high CPU usage. They simply show error message posted above and gracefully exit. I think that the fact they exit with exit code 1 causes that IntelliJ/Cloud code runs this command over and over again

**Comment by ivanporty** [comment]
@bikerp does it reproduce for you with the latest plugin version? We did a minor touch there to ensure we do not read credential too often.

**Comment by bikerp** [comment]
@ivanporty I tried with the latest versions:
IntelliJ IDEA 2020.3.3
Cloud SDK 332.0.0
Cloud Code plugin 21.2.3.202
and the issue still persists

**Comment by ivanporty** [comment]
@bikerp Any chance we could take a look at your log files when you experience this issue please? `IntelliJ Menu -> Help -> Show Log in Finder`

**Comment by bikerp** [comment]
@ivanporty Sure, where should I upload the log files?

**Comment by ivanporty** [comment]
Can you please attach it right here in the issue (drop into comment field) or is it too big?

**Comment by bikerp** [comment]
@ivanporty Here is the log file
[idea.log.gz](https://github.com/GoogleCloudPlatform/cloud-code-intellij/files/6363194/idea.log.gz)

**Comment by ivanporty** [comment]
@bikerp it seems that the issue is somewhat isolates around the way you login from the command line. While Cloud Code is able to read the list of accounts and get some credentials, many calls to APIs fail asking for re-authentication and refresh token seems to be set to happen very frequently. Is there anything very specific about your way of logging in, you mentioned OIDC, it seems `auth describe` and `config config-helper` commands constantly fail. When you login, are you able to run `auth list` command successfully?
